src/core/thread_pool.h

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#pragma once
+
+#include <condition_variable>
+#include <deque>
+#include <future>
+#include <mutex>
+#include <thread>
+#include <vector>
+
+#include "src/core/cuda_utils.h"
+#include "src/core/status.h"
+#include "src/core/sync_queue.h"
+
+namespace nvidia { namespace inferenceserver {
+
+class WorkerThreadPool {
+ public:
+  WorkerThreadPool(int thread_count) : thread_count_(thread_count)
+  {
+    worker_threads_.reserve(thread_count);
+    futures_.reserve(thread_count);
+  }
+
+  ~WorkerThreadPool();
+
+  // Wait for all pending worker threads to finish.
+  Status AwaitCompletion();
+
+  // Get Id of next available worker thread. If all workers are occupied then
+  // wait for them to finish and start from the beginning.
+  Status GetNextAvailableId(int* worker_id);
+
+  // Add CopyBuffer task to queue
+  Status AddTask(
+      const std::string& msg, const TRITONSERVER_MemoryType src_memory_type,
```

Comment:
You should either change the class name to be specific for data copy or, for generality, wrap all the arguments and the `CopyHandler()` below into a lambda that is ready to be called without arguments. You probably need this class to be template class for latter

Reply:
Yes I think it is better to maintain it as a template so that we can do this for other operations that can benefit from being run in parallel.

Reply:
On second thought aggregating the status of CopyBuffer means that the class cannot be generic. I will fix the name to avoid confusion.

Reply:
Why it can't be generic? We can just formulate that the Status object is used to signal whether the task is performed successfully. This class doesn't care about the detail of the Status object anyway, it is left for the caller to interpret the Status object.